File: QuickSort.cpp
#include <vector>

using namespace std;

template <typename T>
void myQuickSort(int l,int r,vector<T> &arr){
	if(l>=r) return;
	T pivot = arr[r];
	int lidx = l, ridx = r-1;
	while(lidx<ridx){
		while(arr[lidx] < pivot && lidx<ridx) lidx++;
		while(arr[ridx] > pivot && lidx<ridx) ridx--;
		if(lidx<ridx){
			T tmp = arr[lidx];
			arr[lidx] = arr[ridx];
			arr[ridx] = tmp;
		}
	}
	if(arr[lidx] > pivot){
		T tmp = arr[lidx];
		arr[lidx] = pivot;
		arr[r] = tmp;
	}
	myQuickSort(l,lidx,arr);
	myQuickSort(ridx+1,r,arr);
}

File: QuickSort/QuickSort.java
package QuickSort;

public class QuickSort {
    public void sort(int[] arr,int l,int r){
        if(l+1 >= r) return;
        int pivot = arr[r-1];
        int idx = l-1;
        for(int i=l;i<r-1;i++){
            if(arr[i] < pivot){
                swap(arr,++idx,i);
            }
        }
        swap(arr,++idx,r-1);
        sort(arr,l,idx);
        sort(arr,idx+1,r);
    }

    private void swap(int[] arr,int a,int b){
        int tmp = arr[b];
        arr[b] = arr[a];
        arr[a] = tmp;
    }
}
